astropy/units/quantity_helper/helpers.py

```diff
@@ -34,7 +35,28 @@ def get_converter(from_unit, to_unit):
     """Like Unit.get_converter, except returns None if no scaling is needed,
     i.e., if the inferred scale is unity.
     """
-    converter = from_unit.get_converter(to_unit)
+    try:
+        converter = from_unit.get_converter(to_unit)
+    except AttributeError as e:
+        # Check for lack of unit only now, to avoid delay for cases where a unit
+        # unit was present. Note that cases where dimensionless is expected are
+        # already short-circuited; here, we cover just the case where, e.g., the
+        # user has done u.add_enabled_equivalencies(u.dimensionless_angles()).
+        if from_unit is not None:  # pragma: no cover
+            raise
+        try:
+            converter = dimensionless_unscaled.get_converter(to_unit)
+        except UnitsError:
+            note = (
+                "Input without a 'unit' attribute? Such input is treated "
+                f"as dimensionless and cannot be converted to {to_unit}."
+            )
+            if sys.version_info >= (3, 11):
+                e.add_note(note)
```

Comment:
Belated comment: I don't know how to feel about having error behavior be different based on Python version. But probably okay for this one case. 🤷‍♀️

Reply:
I did check that the actual error message is exactly identical -- I thought it was easier to do it like this rather than wait for our minimum version of python to be 3.11.

But, actually, aren't we dropping support of python 3.10 by October (following SPEC 0)? If so, isn't this OK regardless?

Reply:
But why is "exotic archs" on Python 3.12 failing? See https://github.com/astropy/astropy/pull/16878#pullrequestreview-2261432822